R/subset.R

```diff
@@ -30,13 +30,18 @@ str_subset <- function(string, pattern, negate = FALSE) {
   check_lengths(string, pattern)
   check_bool(negate)
 
-  switch(type(pattern),
-         empty = no_empty(),
-         bound = no_boundary(),
-         fixed = string[str_detect(string, pattern, negate = negate)],
-         coll  = string[str_detect(string, pattern, negate = negate)],
-         regex = string[str_detect(string, pattern, negate = negate)])
+  if (type(pattern) == "empty") no_empty()
+  if (type(pattern) == "bound") no_boundary()
 
+  idx <- str_detect(string, pattern, negate = negate)
+  # str_detect() returns NA for NAs in string, but str_subset() should drop them
+  idx[is.na(idx)] <- FALSE
+  out <- string[idx]
+  # Work around the fact that as.character() drops names
+  nm <- names(out)
+  out <- as.character(out)
+  names(out) <- nm
```

Comment:
Again, I don't think adding a new behaviour for a literal `NA` is worth complicating the code, and if you're going to convince me it needs to be in a separate PR not as an incidental change in this PR which is about preserving names.